### ISSUE TITLE
Introduce full Reactive support in SCS 1.1

### ISSUE THREAD

**Issue by sdeleuze** [spring-cloud/spring-cloud-stream#458]
Since we are in the process of providing a comprehensive Reactive support in various Spring projects based on [Reactive Streams](http://www.reactive-streams.org/) defacto standard, and supporting both Reactor Core and RxJava API out of the box, I think that would make sense for Spring Cloud Stream  to refactor RxJava processor support before `1.0.0.GA` to a more generic Reactive support that will be suitable for Reactive Streams, Reactor Core 2.5+ ([Flux](http://projectreactor.io/core/docs/api/reactor/core/publisher/Flux.html) and [Mono](http://projectreactor.io/core/docs/api/reactor/core/publisher/Mono.html) types implements Reactive Streams `Publisher`) and RxJava 2.x (it is using different packages and is Reactive Streams compliant). 

RxJava 1.x support would be provided as one of the composition library supported, and @smaldini and I could contribute a draft PR for Reactive Streams generic support (`Publisher<O> process(Publisher<I> input)` processor) based on Reactor Core 2.5+ to validate the design (this PR could be integrated when Reactor Core 2.5 GA will be available maybe it could be available before Spring Cloud Stream `1.0.0.GA` release).

My initial proposal to be discussed is the following:
 - Rename `spring-cloud-stream-rxjava` module to `spring-cloud-stream-reactive`
 - Rename `org.springframework.cloud.stream.annotation.rxjava` to `org.springframework.cloud.stream.annotation.reactive`
 - Make `@EnableRxJavaProcessor` usable for any kind of Reactive library by renaming it to `@EnableReactiveProcessor` (that will also help for the upcoming RxJava 2.x)
 - Clarify, with proper package name, class name or Javadoc, that `RxJavaProcessor` is only for RxJava 1.x

@smaldini and I are available to work on this with you. Any thoughts?

**Comment by aclement** [comment]
Just chipping in that this would be nice to see. I do have a programmable module that we use with Flo that depends on those things ( https://github.com/aclement/programmable-rxjava-processor ) but it can be adapted very quickly to simple name changes.

**Comment by mbogoevici** [comment]
Hi, 

A few thoughts on this:

First, I do appreciate the thought behind @sdeleuze and @smaldini 's proposal - we do need to set up a framework for providing reactive support in Spring Cloud Stream, and it must not be tied to RxJava 1.0.  In fact, I am very keen in making reactive support  a major theme of a SCSt 1.1.0.M1 release (along with Kafka 0.9), and have that follow shortly after 1.0. GA. The goal would be to be ready before Spring IO. 

Second, I do not believe that the current `RxJavaProcessor` is the correct approach on the long run , even if we do provide it currently, with limited support. We do that because Spring XD had a similar feature (where it actually felt very natural), and, at least in the early days of Spring Cloud Stream, we wanted to allow users to migrate their existing code. But, on the long run, it is quite limiting compared to other alternatives (see below), in that it assumes only an input and output channel, does not allow the creation of a reactive sink for example, etc.

Third, one of the considerations behind the late addition of the `@StreamListener` annotation was that it could eventually support reactive types as arguments and return values (with much more flexibility in terms of handling method signature and return parameters, and that includes using `@Input` and `@Output`-annotated parameters to bind arguments to input and output streams, or binding them to return types, e.g. (just for the sake of illustrating a principle)

 ``` 
@StreamListener void join(@Input Publisher<HotDrink> hotDrink, @Input Publisher<ColdDrink> coldDrink, @Output Subscriber<Deliveries> deliveries) {
     return Flux.combineLatest(hotDrink,coldDrink, (h,c)-> new Delivery(h,c)).subscribeWith(deliveries);
}
```

Fourth, one of the reasons for the Binder being parameterized is because we considered the possibility of using reactive types as bound elements, outside `MessageChannel`, e.g `Publisher` and `Subscriber`, and by way of this allowing binder implementations that fit the reactive model better.

In summary, to be honest, I'd much prefer to explore some more flexible extension options, such  as `@StreamListener` and Reactive Type binders and deprecate `EnableRxJava` and `RxJavaProcessor` on the long run, so perhaps instead of retrofitting that API, we could get a head start on a more solid and general approach for a reactive API that could be the basis of 1.1 support. What do you think? I am currently on vacation, but I would like to keep this discussion open.

**Comment by sdeleuze** [comment]
@mbogoevici I fully agree that current `RxJavaProcessor` is just one use case possible among others, and that allowing more powerful and global approach like `@StreamListener` and reactive binders is a much better solution. In fact I created this issue with only short term feedback for 1.0, and planned to create another one with a lot of topics you introduced in your previous comment ;-)

In addition to other Reactive libraries support (including Reactive Streams obviously), I think the approach you described is interesting because it allows to really work at the stream level, not message by message, allowing to managing the demand (backpressure), retry, buffer, etc.

One of the major challenges is to be able to build a full Reactive pipeline in order to avoid loosing all the advantages of real async non-blocking with backpressure pipeline, so that's great for `Binder` to be a parametrized. That could enable us to build Reactive binders (maybe a Kafka one would make sense) and play well with our [Reactive HTTP client](https://github.com/spring-projects/spring-reactive/issues/13) or with [Reactive Spring Data modules](https://github.com/mp911de/spring-data-mongodb/blob/7c758318b09106b4d36bb0d4927606f6ad9f771b/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/ReactiveMongoOperations.java#L71) that could be used for `@StreamListener` implementations.

Thanks for sharing your thoughts, I guess we will continue the discussion after your vacations. In any case what you have in mind for the Spring Cloud Stream 1.1 Reactive support looks promising, and I am sure all the Spring Reactive teams will be happy to help and contribute.

**Comment by bassemZohdy** [comment]
Actually reactive-streams is going to be in Java 9 as Flow.java may have to consider it, cloud-stream or cloud-flow.

**Comment by sdeleuze** [comment]
@bassemZohdy Following Spring Framework 5 move to a Java 8+ baseline, I guess other Spring projects won't be Java 9+ based anytime soon but Java 9 Flow is [already supported out of the box by Reactor Core 2.5](https://github.com/reactor/reactor-core/blob/master/src/main/java/reactor/core/converter/FlowPublisherConverter.java) so this is likely to be supported.

**Comment by smaldini** [comment]
@bassemZohdy We expect to support more Flow overtime as Java 9 gets wider audience, but all current Reactive Streams are based on the JAR itself now. Reactor offers direct support as mentioned by @sdeleuze . The good thing is at least using a library implementation might abstract away the need to do large scale package changes for the user. The contract otherwise beyond package naming is the same and the same engine driven by reactor-core can convert to reactive-streams.jar or Flow.

@mbogoevici I agree with what your said it would prove flexible and elegant. I would add it would be great to refine the output type to dynamic output using a specific `StreamSender` argument input with `StreamSender.send(Publisher) : Mono<Void>` , sending all the passed sequence and returning a completable or errorable result at some point for the sequence, e.g.

```java
@StreamListener void join(@Input Publisher<HotDrink> hotDrink, @Input Publisher<ColdDrink> coldDrink, @Output Subscriber<Deliveries> deliveries, StreamSender out) {
    Flux.combineLatest(hotDrink,coldDrink, (h,c)-> new Delivery(h,c)).subscribeWith(deliveries)
       .window(100, 150) //100 or 150ms timeout
       .consume(out::send);
}
```

**Comment by sdeleuze** [comment]
Like discussed with @mbogoevici, and in line with what @smaldini proposed, it seems we could support Reactive types in Spring Cloud Stream programming model in a similar fashion to what we do in Spring Web Reactive side, leading to a consistent high-level Reactive programming model.

`@StreamListener` would be being similar to `@RequestMapping`, `@Input` + Reactive type like `Publisher`, `Flux` or `Observable` could match with what we do with `@RequestBody` Reactive parameters, and maybe `@Output` (could be optional like we do with `RestControler`) similar to `@ResponseBody` Reactive type return value.

Upcoming [Spring Framework Reactive codecs](https://github.com/spring-projects/spring-reactive/tree/master/src/main/java/org/springframework/core/codec) are something that could make sense since there are intended to be used for any kind of use cases, not just for HTTP (see [CodecHttpMessageConverter](https://github.com/spring-projects/spring-reactive/blob/master/src/main/java/org/springframework/http/converter/reactive/CodecHttpMessageConverter.java) to see how we integrate them to deal with HTTP specific features).
